productivity-suite/piercing-library/src/message-bus.ts

```diff
@@ -0,0 +1,113 @@
+export type MessageHandler = {
+  eventName: string;
+  callback: (detail: any) => void;
+};
+
+export type DispatchOptions = { singleUse?: boolean };
+
+export interface MessageBus {
+  dispatch(eventName: string, detail?: any, options?: DispatchOptions): void;
+  listen(handler: MessageHandler): () => void;
+}
+
+export class BrowserMessageBus implements MessageBus {
+  lastMessagesMap = new Map<
+    string,
+    { detail: any; dispatchOptions?: DispatchOptions } | null
+  >();
+  handlers: MessageHandler[] = [];
+
+  dispatch(eventName: string, detail: any, dispatchOptions: DispatchOptions) {
+    const handlersForEvent = this.handlers.filter(
+      ({ eventName: handlerEventName }) => handlerEventName === eventName
+    );
+    handlersForEvent.forEach(({ callback }) => callback(detail));
+    const eventConsumed = handlersForEvent.length && dispatchOptions?.singleUse;
+    if (!eventConsumed) {
+      this.lastMessagesMap.set(eventName, {
+        detail,
+        ...(dispatchOptions ? { dispatchOptions } : {}),
+      });
+    }
+  }
+
+  listen(handler: MessageHandler) {
+    const lastMessage = this.lastMessagesMap.get(handler.eventName);
+    if (lastMessage) {
+      handler.callback(lastMessage.detail);
+      if (lastMessage.dispatchOptions?.singleUse) {
+        this.lastMessagesMap.set(handler.eventName, null);
+      }
+    }
+    this.handlers.push(handler);
+    return () => {
+      this.handlers = this.handlers.filter((h) => h !== handler);
+    };
+  }
+}
+
+// export class ServerSideMessageBus implements MessageBus {
```

Comment:
all this commented out code is to be introduced for the Server side part of the message bus